embedchain/vectordb/pineconedb.py

```diff
@@ -0,0 +1,180 @@
+                    "config is not a `PineconeDbConfig` instance. "
+                    "Please make sure the type is right and that you are passing an instance."
+                )
+            self.config = config
+        self.client = self._setup_pinecone_index()
+        # Call parent init here because embedder is needed
+        super().__init__(config=self.config)
+
+    def _initialize(self):
+        """
+        This method is needed because `embedder` attribute needs to be set externally before it can be initialized.
+        """
+        if not self.embedder:
+            raise ValueError("Embedder not set. Please set an embedder with `set_embedder` before initialization.")
+
+    # Loads the Pinecone index or creates it if not present.
+    def _setup_pinecone_index(self):
+        pinecone.init(
+            api_key=os.environ.get("PINECONE_API_KEY"),
+            environment=os.environ.get("PINECONE_ENV"),
+        )
+        self.index_name = self._get_index_name()
+        indexes = pinecone.list_indexes()
+        if indexes is None or self.index_name not in indexes:
+            pinecone.create_index(name=self.index_name, metric=self.config.metric, dimension=self.config.dimension)
+        return pinecone.Index(self.index_name)
+
+    def get(self, ids: Optional[List[str]] = None, where: Optional[Dict[str, any]] = None, limit: Optional[int] = None):
+        """
+        Get existing doc ids present in vector database
+
+        :param ids: _list of doc ids to check for existence
+        :type ids: List[str]
+        :param where: to filter data
+        :type where: Dict[str, any]
+        :return: ids
+        :rtype: Set[str]
+        """
+        ids = set()
+        for id in ids:
+            result = self.client.query(ids=id)
+            if result["matches"] is not None and result["matches"] != []:
+                ids.add(id)
+
+        return {"ids": ids}
+
+    def add(
+        self,
+        embeddings: List[List[float]],
+        documents: List[str],
+        metadatas: List[object],
+        ids: List[str],
+        skip_embedding: bool,
+    ):
+        """add data in vector database
+
+        :param documents: list of texts to add
+        :type documents: List[str]
+        :param metadatas: list of metadata associated with docs
+        :type metadatas: List[object]
+        :param ids: ids of docs
+        :type ids: List[str]
+        """
+
+        docs = []
+        if embeddings is None:
+            embeddings = self.embedder.embedding_fn(documents)
+        for id, text, metadata, embedding in zip(ids, documents, metadatas, embeddings):
+            metadata["text"] = copy.copy(text)
+            docs.append(
+                {
+                    "id": id,
+                    "values": embedding,
+                    "metadata": metadata,
+                }
+            )
+
+        for i in range(0, len(docs), self.BATCH_SIZE):
+            self.client.upsert(docs[i : i + self.BATCH_SIZE])
```

Comment:
I am not sure if the .add() is working as intended. I tried to do this but it doesn't work as expected:

```python
In [27]: pinecone_app.query("What is net worth of Elon Musk?")
Out[27]: "As of September 2021, Elon Musk's net worth is estimated to be around $300 billion, making him one of the richest individuals in the world. However, please note that net worth can fluctuate over time due to various factors such as stock market fluctuations and business ventures."

In [28]: pinecone_app.add("https://www.forbes.com/profile/elon-musk")
Out[28]: '8cf46026cabf9b05394a2658bd1fe890'

In [29]: pinecone_app.query("What is net worth of Elon Musk?")
Out[29]: "As of September 2021, Elon Musk's net worth is estimated to be around $300 billion, making him one of the richest individuals in the world. However, please note that net worth can fluctuate over time due to various factors such as stock market fluctuations and business ventures."
```

Since the forbes link contains the latest info, it should give a different answer after adding the web page. 


After looking at the pinecone UI, seems like the same doc gets indexed multiple times so likely an issue with the `.add()` function only. Please see the screenshot below.

<img width="1544" alt="Screenshot 2023-10-09 at 11 47 18 AM" src="https://github.com/embedchain/embedchain/assets/2945708/e944b733-ed96-481f-b38f-bf50fb3240e7">

Reply:
Nice catch, fixed it!